### PR TITLE
enh: lenient url validation

### LINKED ISSUE #6765: Lenient URL validation
URL validation is strict and requires protocol as well. Please make it lenient and add `https://` automatically if not present

### PR DESCRIPTION
**Pull request by maze-runnar**
Fixes #6765 
reference - https://marshmallow.readthedocs.io/en/latest/api_reference.html#marshmallow.fields.Url
and https://marshmallow.readthedocs.io/en/latest/_modules/marshmallow/validate.html#Validator

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by iamareebjamal** [comment]
Read my review

**Comment by maze-runnar** [comment]
@iamareebjamal , these changes are validating URL or not? For being more specific for lanient what should i do next? All i got from docs is -

```
class Url(String):
    """A validated URL field. Validation occurs during both serialization and
    deserialization.

    :param default: Default value for the field if the attribute is not set.
    :param relative: Whether to allow relative URLs.
    :param require_tld: Whether to reject non-FQDN hostnames.
    :param schemes: Valid schemes. By default, ``http``, ``https``,
        ``ftp``, and ``ftps`` are allowed.
    :param kwargs: The same keyword arguments that :class:`String` receives.
    """

    default_error_messages = {"invalid": "Not a valid URL."}

    def __init__(
        self,
        *,
        relative: bool = False,
        schemes: types.StrSequenceOrSet = None,
        require_tld: bool = True,
        **kwargs
    ):
        super().__init__(**kwargs)

        self.relative = relative
        self.require_tld = require_tld
        # Insert validation into self.validators so that multiple errors can be stored.
        original_validators = list(self.validators)
        # FIXME: Why doesn't mypy think validate.URL is a callable here?
        validator = typing.cast(
            typing.Callable[[typing.Any], typing.Any],
            validate.URL(
                relative=self.relative,
                schemes=schemes,
                require_tld=self.require_tld,
                error=self.error_messages["invalid"],
            ),
        )
        self.validators = [validator] + original_validators

```

```fields.Url(*, relative: bool = False, schemes: Union[Sequence[str], Set[str]] = None, require_tld: bool = True, **kwargs)[source]```

**Comment by iamareebjamal** [comment]
Read the issue again. I should be able to post a URL with or without protocol and it should add protocol itself. I never said the solution will be present in docs

**Comment by iamareebjamal** [comment]
It's lenient, not lanient. It's always better to understand the issue before sending a PR

**Comment by maze-runnar** [comment]
`validate=lambda url: url if (url.startswith('https://')) else url.replace('http://', 'https://') if (url.startswith('http://')) else 'https://'+ url`
@iamareebjamal, is it right?

**Comment by maze-runnar** [comment]
> It's lenient, not lanient. 

oh! sorry for this.

**Comment by iamareebjamal** [comment]
Write some tests for this function

**Comment by maze-runnar** [comment]
> Write some tests for this function

```
validate = lambda url: url if (url.startswith('https://')) else url.replace('http://', 'https://') if (url.startswith('http://')) else 'https://'+ url


assert  validate('github.com') == 'https://github.com' #True
assert validate('http://github.com') == 'https://github.com' #True
assert validate('https://github.com') == 'https://github.com' #True
assert validate('linkedin.com/in/') == 'https://linkedin.com/in/'#True
assert validate('www.linkedin.com/in/') == 'https://www.linkedin.com/in/' #True
```

**Comment by iamareebjamal** [comment]
In the PR

**Comment by maze-runnar** [comment]
How to write tests for  validate in `access_url = fields.Url(allow_none=True, validate=lambda url: url if (url.startswith('https://')) else url.replace('http://', 'https://') if (url.startswith('http://')) else 'https://'+ url)
  `
when validate is in itself in fields.Url?
 or do i need to make it like  following - 

```
@validates_schema(pass_original=True)
    def validate_url(self, data, original_data):
        if 'id' in original_data['data']:
            access_code = AccessCode.query.filter_by(id=original_data['data']['id']).one()
        if 'access_url' not in data:
            data['access_url'] == access_code.access_url

        validate = lambda url: url if (url.startswith('https://')) else url.replace('http://', 'https://') if (url.startswith('http://')) else 'https://'+ url
        url = validate(data['access_url'])
        if(not url.startswith("https://")):
            raise UnprocessableEntity({'pointer': '/data/attributes/access_url'},
                                      "put a valid url")
```
And then for test - 
```
def test_lenient_url(self):
        schema = AccessCodeSchema()
        original_data = {
            'data': {}
        }
        data = {
            'access_url': "github.com"
        }
        AccessCodeSchema.validate_url(schema, data, original_data)

``

**Comment by iamareebjamal** [comment]
First you can write unit tests for the function itself like you had written above, and then you can use schema .loads and validate it it transformed to correct URL

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602&bid=16306271)

File: app/api/schema/access_codes.py [hunk 1]
```diff
@@ -73,7 +73,10 @@ def validate_order_quantity(self, data, original_data):
 
     id = fields.Integer(dump_ony=True)
     code = fields.Str(required=True)
-    access_url = fields.Url(allow_none=True)
+    access_url = fields.Url(allow_none=True,
+                            validate=lambda url: url if (url.startswith('https://')) 
+                            else url.replace('http://', 'https://') if (url.startswith('http://'))
```
Comment (niranjan94) [inline app/api/schema/access_codes.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

File: app/api/schema/access_codes.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from marshmallow import validates_schema
+from marshmallow import validates_schema, validate
```
Comment (niranjan94) [inline app/api/schema/access_codes.py]: Codacy Issue found: [Unused validate imported from marshmallow](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

File: app/api/schema/access_codes.py [hunk 3]
```diff
@@ -73,7 +73,10 @@ def validate_order_quantity(self, data, original_data):
 
     id = fields.Integer(dump_ony=True)
     code = fields.Str(required=True)
-    access_url = fields.Url(allow_none=True)
+    access_url = fields.Url(allow_none=True,
+                            validate=lambda url: url if (url.startswith('https://'))
```
Comment (niranjan94) [inline app/api/schema/access_codes.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

File: app/api/schema/access_codes.py (hunk 2 above)
Comment (niranjan94) [inline app/api/schema/access_codes.py]: Codacy Issue found: ['marshmallow.validate' imported but unused](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- tests/all/unit/api/validation/test_access_codes.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602&bid=16306271)

**Comment by maze-runnar** [comment]
@iamareebjamal  do i should make changes like above, for all files?

File: app/api/schema/access_codes.py (hunk 2 above)
Comment (niranjan94) [inline app/api/schema/access_codes.py]: Codacy Issue found: [Unused validate imported from marshmallow](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

Comment (niranjan94) [inline app/api/schema/access_codes.py]: Codacy Issue found: ['marshmallow.validate' imported but unused](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

File: tests/all/unit/api/validation/test_access_codes.py [hunk 4]
```diff
@@ -65,6 +65,14 @@ def test_date_pass(self):
             'valid_till': datetime.datetime(2019, 1, 1),
         }
         AccessCodeSchema.validate_date(schema, data, original_data)
+
```
Comment (niranjan94) [inline tests/all/unit/api/validation/test_access_codes.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

File: app/api/schema/access_codes.py (hunk 3 above)
Comment (niranjan94) [inline app/api/schema/access_codes.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

File: tests/all/unit/api/validation/test_access_codes.py [hunk 5]
```diff
@@ -65,6 +65,14 @@ def test_date_pass(self):
             'valid_till': datetime.datetime(2019, 1, 1),
         }
         AccessCodeSchema.validate_date(schema, data, original_data)
+        
+    def test_url(self):
+        validate = lambda url: url if (url.startswith('https://')) else url.replace('http://', 'https://') if (url.startswith('http://')) else 'https://'+ url
```
Comment (niranjan94) [inline tests/all/unit/api/validation/test_access_codes.py]: Codacy Issue found: [Line too long (158/150)](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

File: app/api/schema/access_codes.py (hunk 1 above)
Comment (niranjan94) [inline app/api/schema/access_codes.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4929602)

**Comment by iamareebjamal** [comment]
No, as written in the review, create a function, and test it. How are you going to test a lambda function?

**Pull request closed without merging**